relay-server/src/actors/upstream.rs

```diff
@@ -276,161 +292,226 @@ impl UpstreamRelay {
+                    .response_sender
                     .send(Err(UpstreamRequestError::BuildFailed(e)))
                     .ok();
-                return Box::new(futures::future::err(()));
+                return;
             }
             Ok(client_request) => client_request,
         };
 
         // we are about to send a HTTP message keep track of requests in flight
         self.num_inflight_requests += 1;
 
-        let future = client_request
+        client_request
             .send()
-            // We currently use the main connection pool size limit to control how many events get
-            // sent out at once, and "queue" up the rest (queueing means that there are a lot of
-            // futures hanging around, waiting for an open connection). We need to adjust this
-            // timeout to prevent the queued events from timing out while waiting for a free
-            // connection in the pool.
-            //
-            // This may not be good enough in the long run. Right now, filling up the "request
-            // queue" means that requests unrelated to `store` (queries, proxied/forwarded requests)
-            // are blocked by store requests. Ideally, those requests would bypass this queue.
-            //
-            // Two options come to mind:
-            //   1. Have own connection pool for `store` requests
-            //   2. Buffer up/queue/synchronize events before creating the request
             .wait_timeout(self.config.event_buffer_expiry())
             .conn_timeout(self.config.http_connection_timeout())
             // This is the timeout after wait + connect.
             .timeout(self.config.http_timeout())
             .track(ctx.address().recipient())
             .map_err(UpstreamRequestError::SendFailed)
             .and_then(handle_response)
-            .then(|x| {
-                response_sender.send(x).ok();
-                Ok(())
-            });
-        Box::new(future)
+            .into_actor(self)
+            .then(|send_result, slf, ctx| {
+                slf.handle_http_response_status(request, send_result, ctx)
+                    .into_actor(slf)
+            })
+            .spawn(ctx);
+    }
+
+    /// Checks what happened to the request and takes appropriate action
+    ///
+    /// - if the request was sent, notify the response sender
+    /// - if the request was not send:
+    ///     - if it was a network error
+    ///         - if request should be retried
+    ///             - schedule a retry
+    ///         - if there is no scheduled Authentication request
+    ///             - schedule authentication (with backoff)
+    ///     - if it was a non recoverable error, notify the response sender with the error
+    fn handle_http_response_status(
+        &mut self,
+        request: UpstreamRequest,
+        send_result: Result<ClientResponse, UpstreamRequestError>,
+        ctx: &mut Context<Self>,
+    ) -> impl Future<Item = (), Error = ()> {
```

Comment:
Please name the future type that is returned from this function.

Reply:
ok